### ISSUE TITLE
rev-parse when using TAGNAME fails

### ISSUE THREAD

**Issue by cliles** [javiplx/jenkins-gitlab-hook-plugin#90]
Following issue #45 from elvanja/jenkins-gitlab-hook-plugin and the documentation -

Setting the branch specifier to ```refs/tags/${TAGNAME}``` works and the build is triggered, but when the build runs, I am getting the following.

```
 > git rev-parse refs/tags/^{commit} # timeout=10
 > git rev-parse refs/remotes/origin/refs/tags/^{commit} # timeout=10
 > git rev-parse refs/tags/^{commit} # timeout=10
ERROR: Couldn't find any revision to build. Verify the repository and branch configuration for this job.
```

It seems that for unknown reason the rev-parse is not picking up the TAGNAME.

**Comment by javiplx** [comment]
Just to be sure, did you set TAGNAME as job parameter? Also, if you see anything relevant to the plugin in the jenkins log, please attach.

**Comment by cliles** [comment]
Yes, TAGNAME is configured as a parameter. 

Log does not show any exceptions, and only shows that the build is scheduled for polling when the web hook fires. It looks like the tag is coming over from gitlab and the plugin is correctly triggering the build that is configured for ```refs/tags/${TAGNAME}```

```
Nov 05, 2015 11:13:51 AM org.jruby.javasupport.JavaMethod invokeDirectWithExceptionHandling
INFO: gitlab web hook triggered for
   - repo url: git@gitlab:/validation.git
   - branch: 
   - with payload:
{
  "object_kind": "tag_push",
  "before": "0000000000000000000000000000000000000000",
  "after": "c1ac8ba6e6f5ca7df6716f320f5509408a26419f",
  "ref": "refs/tags/v0.0.4",
  "checkout_sha": "c1ac8ba6e6f5ca7df6716f320f5509408a26419f",
  "message": null,
  "user_id": 126,
  "user_name": "cliles",
  "user_email": "cliles",
  "project_id": 676,
  "repository": {
    "name": "validation",
    "url": "git@gitlab:emdeonone/validation.git",
    "description": "",
    "homepage": "https://gitlab/validation",
    "git_http_url": "https://gitlab/validation.git",
    "git_ssh_url": "git@gitlab:/validation.git",
    "visibility_level": 0
  },
  "commits": [
    {
      "id": "c1ac8ba6e6f5ca7df6716f320f5509408a26419f",
      "message": "reseting release version\n",
      "timestamp": "2015-11-04T17:03:04-08:00",
      "url": "https://gitlab/validation/commit/c1ac8ba6e6f5ca7df6716f320f5509408a26419f",
      "author": {
        "name": "cliles",
        "email": "cliles"
      }
    }
  ],
  "total_commits_count": 1
}
Nov 05, 2015 11:13:51 AM org.jruby.javasupport.JavaMethod invokeDirectWithExceptionHandling
INFO: matching projects:
   - validation
   - validation-tagbuild
Nov 05, 2015 11:13:51 AM org.jruby.javasupport.JavaMethod invokeDirectWithExceptionHandling
INFO: validation-tagbuild scheduled for polling
Nov 05, 2015 11:13:51 AM hudson.triggers.SCMTrigger$Runner run
INFO: SCM changes detected in validation-tagbuild. Triggering  #10
```

job xml is

```
<?xml version='1.0' encoding='UTF-8'?>
<project>
  <actions/>
  <description></description>
  <logRotator class="hudson.tasks.LogRotator">
    <daysToKeep>30</daysToKeep>
    <numToKeep>-1</numToKeep>
    <artifactDaysToKeep>-1</artifactDaysToKeep>
    <artifactNumToKeep>-1</artifactNumToKeep>
  </logRotator>
  <keepDependencies>false</keepDependencies>
  <properties>
    <hudson.model.ParametersDefinitionProperty>
      <parameterDefinitions>
        <hudson.model.StringParameterDefinition>
          <name>TAGNAME</name>
          <description></description>
          <defaultValue></defaultValue>
        </hudson.model.StringParameterDefinition>
      </parameterDefinitions>
    </hudson.model.ParametersDefinitionProperty>
  </properties>
  <scm class="hudson.plugins.git.GitSCM" plugin="git@2.4.0">
    <configVersion>2</configVersion>
    <userRemoteConfigs>
      <hudson.plugins.git.UserRemoteConfig>
        <url>git@gitlab:/validation.git</url>
        <credentialsId>XXX</credentialsId>
      </hudson.plugins.git.UserRemoteConfig>
    </userRemoteConfigs>
    <branches>
      <hudson.plugins.git.BranchSpec>
        <name>refs/tags/${TAGNAME}</name>
      </hudson.plugins.git.BranchSpec>
    </branches>
    <doGenerateSubmoduleConfigurations>false</doGenerateSubmoduleConfigurations>
    <browser class="hudson.plugins.git.browser.GitLab">
      <url>https://gitlab</url>
      <version>7.11</version>
    </browser>
    <submoduleCfg class="list"/>
    <extensions/>
  </scm>
  <canRoam>true</canRoam>
  <disabled>false</disabled>
  <blockBuildWhenDownstreamBuilding>false</blockBuildWhenDownstreamBuilding>
  <blockBuildWhenUpstreamBuilding>false</blockBuildWhenUpstreamBuilding>
  <triggers>
    <hudson.triggers.SCMTrigger>
      <spec></spec>
      <ignorePostCommitHooks>false</ignorePostCommitHooks>
    </hudson.triggers.SCMTrigger>
  </triggers>
  <concurrentBuild>false</concurrentBuild>
  <builders>
    <hudson.tasks.Shell>
      <command>bundle install</command>
    </hudson.tasks.Shell>
    <hudson.plugins.rake.Rake plugin="rake@1.8.0">
      <rakeInstallation>(Default)</rakeInstallation>
      <rakeFile></rakeFile>
      <rakeLibDir></rakeLibDir>
      <rakeWorkingDir></rakeWorkingDir>
      <tasks>test</tasks>
      <silent>false</silent>
      <bundleExec>true</bundleExec>
    </hudson.plugins.rake.Rake>
  </builders>
  <publishers>
    <hudson.plugins.cloneworkspace.CloneWorkspacePublisher plugin="clone-workspace-scm@0.6">
      <workspaceGlob>**</workspaceGlob>
      <criteria>Any</criteria>
      <archiveMethod>TAR</archiveMethod>
      <overrideDefaultExcludes>false</overrideDefaultExcludes>
    </hudson.plugins.cloneworkspace.CloneWorkspacePublisher>
  </publishers>
  <buildWrappers>
    <hudson.plugins.ws__cleanup.PreBuildCleanup plugin="ws-cleanup@0.28">
      <deleteDirs>false</deleteDirs>
      <cleanupParameter></cleanupParameter>
      <externalDelete></externalDelete>
    </hudson.plugins.ws__cleanup.PreBuildCleanup>
    <hudson.plugins.ansicolor.AnsiColorBuildWrapper plugin="ansicolor@0.4.2">
      <colorMapName>xterm</colorMapName>
    </hudson.plugins.ansicolor.AnsiColorBuildWrapper>
    <com.cloudbees.jenkins.plugins.sshagent.SSHAgentBuildWrapper plugin="ssh-agent@1.8">
      <credentialIds>
        <string>XX</string>
      </credentialIds>
      <ignoreMissing>false</ignoreMissing>
    </com.cloudbees.jenkins.plugins.sshagent.SSHAgentBuildWrapper>
  </buildWrappers>
</project>
```

**Comment by javiplx** [comment]
As the error seems to have from the git plugin execution, I'll rerun acceptance test with latest versions of required plugins.
The only idea I have is try to build without the workspace clone, although I don't really believe it will change anything.

**Comment by cliles** [comment]
Removing workspace clone does not change the outcome.

My environment is
Jenkins: 1.632
GIT client plugin: 1.19.0
GIT plugin: 2.4.0
Gitlab hook plugin: 1.4.0

**Comment by javiplx** [comment]
It looks like is an issue related with the jenkins version. Most acceptance test are broken on 1.632 due to project creation failure. And although the test for tag building itself does succeed, I believe that is because the point where it is actually broken is not hitted by the test. 
I believe that downgrading Jenkins is probably not an option, because tests also fails with latest LTS (1.625.1)

**Comment by javiplx** [comment]
I have fixed issues related to acceptance tests, which are OK for all recent LTS releases. I've also added an rspec example with your payload, and it is extracted properly, so I'm without any ideas. Which git version are you using? And might you by chance running this on windows? (I've seen some issues with git on windows).
Besides this questions, key issue is to check content of _Git Build Data_ and whether the tag is properly set on _Parameters_.
And the only other approach is to verify/reproduce with command line from the jenkins work directory, issuing the commands seen on the job log to check their output. I expect the tag to be present there, but not sure if it is actually should.

**Comment by javiplx** [comment]
When trying to reproduce, I've found an [already solved (but unreleased) bug](https://issues.jenkins-ci.org/browse/JENKINS-28452). I'm not sure if it could be related to your problem, but it might be worthy to give a chance to the latest snapshot, that you can find at http://maven.jenkins-ci.org/content/repositories/snapshots/org/jenkins-ci/ruby-plugins/gitlab-hook/1.4.1-SNAPSHOT/

**Comment by javiplx** [comment]
Although still unable to reproduce, repeating tests with various plugin versions, I've found a change of behaviour in git plugin, likely related to [a resolved bug _'Branch Specifier regex fails - passed to git rev-parse?'_](https://issues.jenkins-ci.org/browse/JENKINS-26842). I don't see where the relation with your issue could exactly arise, but refspec comparison is geep in the core of gitlab-hook plugin, so maybe downgrading git plugin to 2.3.4 could serve as workaround for your issue. The git-client plugin versión seems not relevant, so you can keep 1.19.0.

**Comment by javiplx** [comment]
I've realized that there is a case where what you describe could happen. If you have a single jenkins project tracking a repository, and that project is prepared to build tags, a standard push will not automatically create a project to track standard branches. Then **_if_** (I'm unsure on that) the jenkins project is built after a push, will not be able to assign TAGNAME because there is no tag on payload, and will produce the job log that you posted.
Now that I'm realized of this case, will try to add to acceptance checks to verify whether it is actually happening, and then try to fix after reproducing it.

**Comment by cliles** [comment]
Webhook is set to trigger on Tag Push Events, the job is configure to only build on tags. To elaborate on your scenario, I'm using the puppet-blacksmith gem to `module:release` my project, which I believe is the equivalent of `git push && git push --tags`.

**Comment by cliles** [comment]
Now I am unable to reproduce as well... let me try again and see, if not I'll close it out.

**Comment by cliles** [comment]
Unable to reproduce :(
